ledger/participant-integration-api/src/main/scala/platform/apiserver/services/ApiCommandCompletionService.scala

```diff
@@ -52,27 +51,23 @@ private[apiserver] final class ApiCommandCompletionService private (
 
         completionsService
           .getCompletions(offset, request.applicationId, request.parties)
-          .via(logger.debugStream(completionsLoggable))
+          .via(logger.debugStream(completionsLogLine))
           .via(logger.logErrorsOnStream)
           .via(StreamMetrics.countElements(metrics.daml.lapi.streams.completions))
     }
 
-  private def completionsLoggable(response: CompletionStreamResponse): String =
-    s"Responding with completions: ${response.completions.toList
-      .map(c => singleCompletionLoggable(c.commandId, c.status.map(_.code)))}"
-
-  private def singleCompletionLoggable(
-      commandId: String,
-      statusCode: Option[Int],
-  ): LoggingEntries =
-    LoggingEntries(
-      logging.commandId(commandId),
-      "statusCode" -> statusCode.fold("")(_.toString),
-    )
-
   override def getLedgerEnd(ledgerId: domain.LedgerId): Future[LedgerOffset.Absolute] =
     completionsService.currentLedgerEnd().andThen(logger.logErrorsOnCall[LedgerOffset.Absolute])
 
+  private def completionsLogLine(response: CompletionStreamResponse): String = {
+    val completionsString = response.completions.view
+      .map(c =>
+        s"{commandId: ${c.commandId}, statusCode: ${c.status.fold("none")(_.code.toString)}}"
+      )
+      .mkString("[", ", ", "]")
+    s"Responding with completions: $completionsString"
+  }
+
```

Comment:
I've seen `Show` from scalaz being used in a few places, any reason not to use it here? Especially as I've see a few `.mkString("[", ", ", "]")` for collections

Reply:
Scalaz is used in a few places but I don't think we use it much in this module. I think in this case, we'd end up defining an orphan instance for `Show[CompletionStreamResponse]` which would be very specific to this particular log line; I'm not sure it's general enough to warrant a typeclass instance like that.

Reply:
Fair enough.
I would normally still add an instance, even if it would be an orphan, can even be scoped locally, just so I don't have mentally process the call to print

Reply:
Well, if it's an orphan instance that's used once… that's the same as a function. Would it help if I pulled out a function here?

I think you make a good point about the `.mkString` call, too. Perhaps this stuff would be better in the logging context. Then we can use `ToLoggingContext`.

Reply:
:+1: for putting these into the logging context.

Reply:
Done. I avoided implementing `ToLoggingValue[CompletionStreamResponse]` because it seemed a bit too niche, as I said above, but I think this code is a lot cleaner.